.husky/pre-commit

```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env sh
+SRC=src/_locales/en/messages.yml
+DST=src/types.generated.d.ts
+
+if git diff --cached -- "$SRC"; then
+  . "$(dirname -- "$0")/_/husky.sh"
+  node scripts/generate-types.js "$SRC" "$DST"
```

Comment:
How about running this in npm scripts `prepare` and `i18n` and ignore `types.generated.d.ts` in the repo?

Reply:
Why? The point of this PR is that changing `messages.yml` also changes `types.generated.d.ts` in the same commit.

Reply:
I also found a problem: it should run in `post-rewrite` and `post-merge`, but in these cases it should probably create a new commit instead of amending the merge...

Reply:
Since it can be generated on demand, why do we keep it in the repo and maintain such a file manually?

Reply:
How else can we make IDE see our message ids?

Reply:
Here is what I suggest:

- Add `src/types.generated.d.ts` to `.gitignore` so it won't be kept in the repo, but the file can still be there
- Instead of running `scripts/generate-types.js` on commit, I suggest running in the following cases:
  - `yarn prepare` - when we install/update deps
  - `yarn i18n` - when we add new i18n keys
  - `post-checkout` - when we switch to a new branch
- For VSCode we can even [add a hook](https://www.roboleary.net/vscode/2020/10/19/vscode-task-onstartup.html) to always recreate the `.d.ts` file on load

BTW I'm a bit confused why we need the hints since we usually add `i18n('xxx')` first and then generate its translation with `yarn i18n`. Is it for reusing existing keys?

Reply:
1. Can you add a locally generated file to jsconfig.json so that the IDE sees it?
2. What about updating the types to match messages.yml while working on a single branch?
3. I've never used `yarn i18n`, do I have to do it each time I edit `en/messages.yml` and if yes, then why don't we auto-run this task via hooks?

>Is it for reusing existing keys?

Yes. Without autocomplete I have to open the file and search inside and that's highly inconvenient due to the presence of descriptions and the structural keys (description and message) which produce false matches while using incremental search or filter. The lack of such autocomplete might be the reason why there are many duplicates in our messages.

Reply:
> 1\. Can you add a locally generated file to jsconfig.json so that the IDE sees it?

Yes, as long as the file exists.

> 2\. What about updating the types to match messages.yml while working on a single branch?

How about running it in `yarn i18n` and adding `yarn i18n` to pre-commit hook?

> 3\. I've never used `yarn i18n`, do I have to do it each time I edit `en/messages.yml`

I think so, actually `en/messages.yml` is generated/updated by `yarn i18n` automatically, adding new keys and removing unused keys (only in prod mode).

Reply:
So we need to commit an empty `types.generated.d.ts` in the repo in order to make jsconfig.json correct? Anyway, the solution that relies on an ignored file seems inherently fragile.

As for `yarn i18n` **current** behavior, why do I need to run it if I've just added a new entry to messages.yml and used it in the code? As for making it run automatically, pre-commit hook is probably insufficient, we'd also need post-merge and post-rebase.

Reply:
> So we need to commit an empty `types.generated.d.ts` in the repo in order to make jsconfig.json correct?

No, it is ignored if the file doesn't exist, and picked by tsserver if the file is present. It's just a match pattern instead of an exact file to be imported.

> Anyway, the solution that relies on an ignored file seems inherently fragile.

Check `node_modules/.vite` for Vite, and `.svelte-kit` for Svelte, they are exactly for the same purpose, which contain types generated on demand, and ignored in git. I think this is quite a common use case for TS/JS.

> why do I need to run it if I've just added a new entry to messages.yml and used it in the code

Firstly, `yarn i18n` will add the keys automatically which might be helpful if you create a new view with many new keys. Secondly, it verifies and formats the message.yml files just like `lint`.

> we'd also need post-merge and post-rebase.

I think we can just hook `prepare` and `post-checkout`, and let the developer deal with the edge cases.

Reply:
> picked by tsserver if the file is present

I see.

> I think we can just hook prepare and post-checkout, and let the developer deal with the edge cases.

Which edge cases? Merging and rebasing is not an edge case. The point of my PR is to make it as reliable as possible. Ideally it should work on-the-fly, but that's only possible via an IDE plugin, so at the least it should regenerate the types when the change is committed in git.

Reply:
BTW I've just tried `gulp i18n` and I really don't like how it modifies 30 files by adding meaningless empty translations for the id I've just added. It bloats my simple commit into a monster in which I can't even see what's changed at a glance.

Reply:
We can change to only modify `en`, and leave others to the transifex script.